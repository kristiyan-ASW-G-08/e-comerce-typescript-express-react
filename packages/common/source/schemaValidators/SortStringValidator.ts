import * as yup from 'yup';

interface SortStringValidatorType {
  sort: 'top' | 'trending' | 'new' | 'replies';
}

const SortStringValidator = yup.object<SortStringValidatorType>().shape({
  sort: yup
    .string()
    .trim()
    .oneOf(['top', 'trending', 'new', 'replies']),
  limit: yup
    .number()
    .integer()
    .min(1)
    .max(50),
  page: yup
    .number()
    .min(1)
    .integer(),
});

export default SortStringValidator;
