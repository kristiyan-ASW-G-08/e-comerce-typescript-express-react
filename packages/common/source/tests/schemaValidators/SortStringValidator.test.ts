import SortStringValidator from '@schemaValidators/SortStringValidator';

describe('SortStringValidator', (): void => {
  const validQueries = [
    {
      limit: 25,
      page: 1,
      sort: 'top',
    },
    {
      limit: 5,
      page: 1,
      sort: 'new',
    },
    {
      limit: 15,
      page: 1,
      sort: 'trending',
    },
    {
      limit: 50,
      page: 100,
      sort: 'replies',
    },
  ];
  const invalidQueries = [
    {
      limit: 100000,
      page: -2,
      sort: 'other',
    },
    {
      limit: -100,
      page: 1,
      sort: 'some',
    },
    {
      limit: 100000000000,
      page: -100,
      sort: 'zeta',
    },
    {
      limit: 50,
      page: 100,
      sort: 'beta',
    },
  ];
  it.each(validQueries)('should validate successfully', async query => {
    await expect(
      SortStringValidator.validate(query, { abortEarly: false }),
    ).resolves.toBe(query);
  });
  it.each(invalidQueries)('should validate unsuccessfully', async query => {
    await expect(
      SortStringValidator.validate(query, { abortEarly: false }),
    ).rejects.toMatchSnapshot();
  });
});
