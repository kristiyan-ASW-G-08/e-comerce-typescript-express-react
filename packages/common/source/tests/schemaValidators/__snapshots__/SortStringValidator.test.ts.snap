// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortStringValidator should validate unsuccessfully 1`] = `[ValidationError: 3 errors occurred]`;

exports[`SortStringValidator should validate unsuccessfully 2`] = `[ValidationError: 2 errors occurred]`;

exports[`SortStringValidator should validate unsuccessfully 3`] = `[ValidationError: 4 errors occurred]`;

exports[`SortStringValidator should validate unsuccessfully 4`] = `[ValidationError: sort must be one of the following values: top, trending, new, replies]`;
